### DISCUSSION TITLE
Devtronic SmartPendant - a way to control grblHAL based CNC machine without PC

### DISCUSSION THREAD

**Discussion by nickshl** [grblHAL/core#465] [Show and tell]
Hi Everyone!

This is the demo of my SmartPendant: https://youtu.be/z-IoQAUr2qA

![Devtronic_SmartPendant_Case](https://github.com/grblHAL/core/assets/38679154/de51cfd6-ba29-4e10-8a83-bfdfb5a8783d)

Source code can be found there: https://github.com/nickshl/SmartPendant

And board schematic along with files for case can be found there: https://github.com/Devtronic-US/SmartPendant

![Devtronic_SmartPendant_1](https://github.com/grblHAL/core/assets/38679154/6e41c999-7a64-4bf7-be8c-99142a7c9758)
![Devtronic_SmartPendant_2](https://github.com/grblHAL/core/assets/38679154/073feb4a-be2a-4edd-843f-00f00ddd3074)

**Comment by Jay-Tech** [comment]
very nice,  alot of features packed in there.  You could get alot of jobs done with out needing a sender app.

**Comment by nickshl** [comment]
@terjeio, there is a bug on "Probe" screen - it may move spindle down instead up after probing. It fixed in source code on GitHub, please update it before try.

**Comment by phil-barrett** [comment]
How does it connect to grblHAL?  I see the UART - is that it?  Does it use the cable for comm as well as power?  If I wanted to support this with my breakout boards what would I need to do?  How much current does it draw? What is the purpose of the MPG_EN pin?

**Comment by nickshl** [comment]
> How does it connect to grblHAL? I see the UART - is that it? Does it use the cable for comm as well as power?

Yes, that is it. GND, UART TX, UART RX and +5V. One cable holds power and data.

> What is the purpose of the MPG_EN pin?

[MPG & DRO mode](https://github.com/grblHAL/core/wiki/MPG-and-DRO-interfaces) -> Mode switching: "_Mode switching is either by signalling via a **dedicated input** on the controller or via a real-time command character 0x8B_".

 MPG_EN is dedicated control request line that can be used to request control if controller has it. Otherwise real time command can be used to request control.

> How much current does it draw?

About 200 mA.

> If I wanted to support this with my breakout boards what would I need to do?

Two UART(TX and RX) lines and power(GND and +5V). There also some settings on grblHAL side needed, but you better ask @terjeio about it. For my board I have to select "_MPG & DRO mode: Test 2_" on Plugin page in Web Builder.

**Reply by phil-barrett** [reply]
OK, so it should work with either of my boards. Any idea when you will have them for sale?

**Reply by nickshl** [reply]
> when you will have them for sale?

Technically I already have. My target price is $100 not included shipping or taxes. At this point it easier for me to work directly with everyone who is interested, since demand isn't seems very high - I didn't sold a single board of [my CNC controller](https://github.com/grblHAL/core/discussions/406) within a month(_which probably a good thing_ 😅). 

I have two fully assembled boards and need to print case for them(_which takes day or two_). I also can build three more, but for those I have to order displays and hand wheels. And I can always order more if there will be demand.

**Reply by phil-barrett** [reply]
OK.  If anyone is interested in something like that, I will send them your way.

**Comment by nickshl** [comment]
@phil-barrett, I decided answered there to this post: https://github.com/grblHAL/core/discussions/462#discussioncomment-8859154

> If you can hit $100 for a complete version of your smartpendant, then it's a very good deal. Though, it's not clear what extras are needed - honestly I am a bit skeptical of the end price. But, it's not all bad.

Skeptical too much or too little? Or if I will be able to deliver for this price? All components and materials cost about $50. So, 100% "profit" pretty damn good. But "profit" in a quotes for a reason: 

> I do see some manufacturing complexity to create a complete unit ready to sell

There is some complexity. JLCPCB will not solder BlackPill, so some soldering needed even if they will place all other components. Then boards have to be washed in isopropyl alcohol, rinsed in water and dried with shopvac in blow mode. I also used trough hole components because of space(_connector for BlackPill_) and robustness(_buttons, SMD can rip of traces if press too hard_) and because of that all this protruding pins have to be milled out to 0.3-0.5 mm height to install display. At this point I do it manually, but in future I plan to make a jig to clamp it easily on a mill and some software to use 3D touch probe to measure height map and mill everything automatically. Then VHB tape(_0.7 mm thickness_) have to be applied to board, kapton tape have to be applied to back of the display(_additional protection to prevent any possible shorts_) and both can be glued together. Sounds as a lot of work, but for bathes of some number of boards shouldn't be too bad.

Then case have to be 3D printed(_I print cases from PLA - may be not a best choice, but easiest to print_) and board should be fit into it(_some mechanical work needed too, but much less than my first prototypes_). I probably will sell those boards(_including handwheel_) without case cheaper at some point for customers willing to print their cases themselves and fit board into it.

**Comment by Jay-Tech** [comment]
Not really an easy post to follow but if you are actively selling these I will buy one.   Got a project I am starting sometime soon  that I would like to usse this for.

**Reply by nickshl** [reply]
> Not really an easy post to follow

As you probably already guessed, I am not an native speaker. We moved to US about 8 years in my 30s and all my English skills I picked up by watching TV and conversations with other people(_which works good for kids, but not so good for adults_)... but there one thing: people so polite there, they afraid to offend anyone, as result they never ever corrects any mistakes! How I suppose to learn language if no one points to my mistakes and tells me how to speak right? 😅

> if you are actively selling these I will buy one

Send me an email - you can use one in YouTube channel description.

There couple I should finish this weekend:
![IMG_20240321_232259058~2](https://github.com/grblHAL/core/assets/38679154/ab8b73d2-85b8-4a77-bea3-867c7a33abd5)
Not a glossy like in a video(_I used textured PEI to print those_) and you can see two bottom buttons(_reset and boot for firmware update_).

**Comment by andrewmarles** [comment]
Have you given any thought to releasing the firmware under an open-source license so that others can use it and help you grow the project as a whole?

**Reply by nickshl** [reply]
Some of the files have license in header. I probably should pay more attention to it. I am completely fine to use it for any non-commercial purposes 🙂

**Reply by andrewmarles** [reply]
No worries.  You definitely want to be more clear that you intend it to be proprietary.  The only licenses that I could find in there was in the FreeRTOS code which is MIT license and GPL on the code from GRBLHAL, both of which are open-source licenses.

**Reply by nickshl** [reply]
>  The only licenses that I could find in there was in the FreeRTOS code which is MIT license and GPL on the code from GRBLHAL

Header in this file: https://github.com/nickshl/SmartPendant/blob/main/DevCore/DevCfg.h
And all header files in DevCore folder has same section with BSD license I believe. Some files in Application folder also has it. As for others... probably just didn't copy-paste it.

> You definitely want to be more clear that you intend it to be proprietary.

I wouldn't call it "proprietary". Open-source licensing has one huge flaw: double taxation. If you makes commercial product, and pays for license, all those expenses can be counted to offset your tax base and only software author will pay taxes for this portion. If you use open source software, it is free, you expenses is 0, and if you decide to make a donation to software author, it can be done only with your profit money you already payed taxes for. But your donation will be counted as income and software author must pay taxes for it too. How do I know? I googled it once to figure out how it works and found this article: https://dev.to/voxpelli/open-source-and-taxes-a-swedish-perspective-gkn . It is about Sweden, not Norway, but close 🙂

**Comment by Jay-Tech** [comment]
> > Not really an easy post to follow

Sorry, I was meaning the questions, commments, and answers jumping into another post then back to this one.  Not  a comment relating to grammer or language , 

> As you probably already guessed, I am not an native speaker. We moved to US about 8 years in my 30s and all my English skills I picked up by watching TV and conversations with other people(_which works good for kids, but not so good for adults_)... but there one thing: people so polite there, they afraid to offend anyone, as result they never ever corrects any mistakes! How I suppose to learn language if no one points to my mistakes and tells me how to speak right? 😅

Your doing  fine. 

> Send me an email - you can use one in YouTube channel description.

Thanks its sent

**Reply by nickshl** [reply]
> Sorry, I was meaning the questions, commments, and answers jumping into another post then back to this one. Not a comment relating to grammer or language ,

> Your doing fine.

Perfect proof of my point! 🤣

**P.S.** obviously it is a joke, no need to continue this further 🙂

**Comment by nickshl** [comment]
The project now fully open source including the PCB: [https://oshwlab.com/nick_shl/smart_pendant](https://oshwlab.com/nick_shl/smart_pendant)

**Reply by mintracer** [reply]
amazing Work! Thx!

**Comment by nickshl** [comment]
A new video about SmartPendant compatibility:

[![IMAGE ALT TEXT HERE](https://img.youtube.com/vi/ToWwejkP6e4/0.jpg)](https://www.youtube.com/watch?v=ToWwejkP6e4)

@phil-barrett I finally redesigned SmartPendant PCB to make it way easily to assemble and plan to sell kit version of SmartPendant withing a couple of weeks. You cold check the video to figure out if it will be compatible with your controller.

**Reply by phil-barrett** [reply]
Sorry, I didn't see the notification.  I was a bit preoccupied after my knee operation.  Anyway, Both the T41U5XBB (iMXRT1062 Teensy) and PicoCNC (RP Pico) meet your 3 requirements.

If you want, I could send you a PicoCNC board to test with.

**Comment by nickshl** [comment]
[![IMAGE ALT TEXT HERE](https://img.youtube.com/vi/WytGEaL6THc/0.jpg)](https://www.youtube.com/watch?v=WytGEaL6THc)

**Reply by phil-barrett** [reply]
That is a helpful video. Hopefully it will encourage sales of your kits.

A couple of thoughts (please don't see this as criticism, I'm just drawing from my experience):
- Soldering SMDs seems to be a sticking point for a lot of potential customers. Any way to get them preinstalled would be a win.
- Loading firmware is another issue for many customers.  I love the Pico approach. Hold the run button while plugging in the USB cable and a file folder opens on the PC (Mac, too). Drag and drop the firmware on the folder and it is done.  Any way to do something similar?
- You should be able to find shorter surface mount pin headers to avoid having the user cut them off.
- You might think about integrating the STM32F onto the PCB and skip the Black Pill entirely.

**Reply by silver2row** [reply]
That is neat. I hope you guys can configure something with pendants and GrblHAL.

Seth

**Reply by nickshl** [reply]
> Soldering SMDs seems to be a sticking point for a lot of potential customers. Any way to get them preinstalled would be a win.

Those boards assembled by JLCPCB. They have two options for PCB assembly: Economy and Standard. For Economy only one side of PCB can be selected for placing parts. The price difference for 10 PCB order is $93.63 or $9.36 per board, which is kind of high considering $60 price of the kit delivered within US. I placed as much as possible parts on one side. Only 4 LEDs and two buttons have to be on the other side for obvious reason. LEDs, as I mention in the video, is nice to have, but not really necessary. Buttons is big enough and easy to solder. I would prefer to use trough-hole buttons, but those buttons have to be very slim and slim button exists only as SMD. Anyway, considered demand on this product, which I can count how many I sold on fingers... using one hand... I more than willing to solder LEDs and buttons myself if customer ask about it.

But, you made a point. I think in addition to kit I can sell fully soldered PCB(_LEDs, Buttons and BlackPill_) with glued display for additional $10-$15 or so.

> Loading firmware is another issue for many customers.  I love the Pico approach. Hold the run button while plugging in the USB cable and a file folder opens on the PC (Mac, too). Drag and drop the firmware on the folder and it is done.  Any way to do something similar?

I don't tried to find solution like that. I think those exists, but to do so, USB bootloader have to be load in STM32 MCU. This bootloader will occupy some flash. Also firmware that will be loaded with this bootloader must be modified and will not run without such bootloader. 
ST Microelectronics included bootloader into STM32. This bootloader can be activated by holding boot button during connection of USB cable(_if controller isn't powered_) or holding boot button while clicking reset button. But instead using Mass Storage Class, ST Microelectronic decided to use DFU. That is why STM32CubeProgrammer should be used to upload firmware. I don't think it is big deal, just a minor inconvenience.

> You might think about integrating the STM32F onto the PCB and skip the Black Pill entirely.

Doesn't make much sense from cost standpoint. STM32F411CEU6 cost about $1 less than BlackPill. Additionally some placing fees and some other components and price will be about the same. I would like to keep this project closer to DIY. In fact whole project can be assembled without fancy purple PCB using some prototyping PCB, buttons and 3 major components: [WeAct BlackPill F411 25M HSE](https://s.click.aliexpress.com/e/_DC6TlGd), [3.5" Display with touchscreen based on ILI9488 LCD controller and FT6236 touch controller](https://www.aliexpress.us/item/3256804935586911.html) and [60 mm 6 pin 100 PPR handwheel](https://s.click.aliexpress.com/e/_DCFuJHr).

<details><summary>That is how I started this project...</summary>
Handwheel went to another device when I received PCB.

![image](https://github.com/user-attachments/assets/9b348d9b-c2b7-4212-98dd-fcdbaaf57bc9)

![image](https://github.com/user-attachments/assets/446ef075-5ee8-460a-b0a0-9e449141ab84)
</details>

**Reply by phil-barrett** [reply]
I've developed a model of DIYers.  There are the hard core ones that can do it all but they represent maybe 5%. The next group are versed in electronics, know how to source parts, can solder and use Ohms law, ...  Maybe 25%.  The next group can solder but have  issues with SMDs, sourcing is hard for them - >50%.  The rest are good at snapping parts together but soldering is intimidating.

Probably the biggest roadblock for customers I have seen is ordering all the different pieces. I estimate that less than 50% of the bare boards ordered ever get put into a machine. Sales really took off when I added the kits. If you can put a full kit together, that drastically increases the chances that someone will get it working and increases your orders. 

The only reason I don't carry a Teensy or Pico is I would have to charge more than they can get elsewhere - not a lot of discount on those, even in decent volume. Making 5% on a Teensy is just not enough to make up for the inevitable losses.

**Comment by nickshl** [comment]
Assembled version [available now](https://devtronic.square.site/product/devtronic-smartpendant-kit-assembled-/4):

![Assembled_Kit_3](https://github.com/user-attachments/assets/c9444e68-73af-4914-bae3-0f18ae7d2c89)
![Assembled_Kit_1](https://github.com/user-attachments/assets/fb154a8c-b88d-4427-991a-e524d34329b5)
![Assembled_Kit_2](https://github.com/user-attachments/assets/83e172a9-eeec-489f-822d-626c055adeeb)

**Comment by nickshl** [comment]
Hi Everyone! Big announcement today: all code generators are removed from the latest firmware! It freed up some RAM to lad programs.
But wait! That is not all! Now possible to use C-like scripts to generate G-Code!
[![IMAGE ALT TEXT HERE](https://img.youtube.com/vi/3LJ1njGXP9w/0.jpg)](https://www.youtube.com/watch?v=3LJ1njGXP9w)
Examples of those scripts can be found there: https://github.com/nickshl/SmartPendant/tree/main/Scripts

@terjeio, just wondering if you used SmartPendant at all and if you have any feedback?

**Comment by nickshl** [comment]
Latest PCB files for board with BlackPull added to GitHub: https://github.com/Devtronic-US/SmartPendant
Order form your favorite PCB supplier and enjoy! Some assembly required. But it isn't that hard as previous available version.

[![YouTube video](https://img.youtube.com/vi/NRwtdEsotCc/0.jpg)](https://www.youtube.com/watch?v=NRwtdEsotCc)
